### PR TITLE
chore: Streamline development experience

### LINKED ISSUE #7: Document how to build project and work with Nix
### Problem

Users coming into the project are expected to know Nix to build the project. If we want people to develop the same way as us, we should add a "developing with nix" section to the readme.

### Proposed solution

Add a section that explains setting up nix and using it for development.

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by phated**
Closes #7

This adopts [direnv](https://direnv.net/) to setup the developers shell and editor environment based on our Nix flake. I've also made some quality-of-life improvements to the flake and added flake-compat files for users without the nix command or flakes enabled (most of that comes from the setup in https://github.com/direnv/direnv-vscode). I've also added some vscode settings and recommended plugins with `direnv-vscode` being the most important; however, the direnv experience gets a little better when the release the feature I submitted in https://github.com/direnv/direnv-vscode/pull/458.

We still need to test this setup in a remote-connection environment and in WSL, but I'm looking to @kobyhallx to test those workflows.

We'll probably want some of the noir language devs to go through my instructions to see if they encounter any problems.

### CODE REVIEW

**Comment by TomAFrench** [comment]
Just to confirm I understand this PR.

Currently in order to build `barretenberg-sys`, developers either need to either:
a) have a globally installed version of barretenberg (requiring them to go off and clone, install deps and build that repo)
b) have nix installed, run `nix develop` to open a new shell which builds the barretenberg dependency and then do all of their work inside that shell (remembering to reopen that shell every time they open the repo)

After this PR we're going to be funnelling people down route B but with `direnv` we can automatically inject the flake's build environment into the user's main shell, making it roughly equivalent to the shell provided by `nix-shell`. Is that correct?

In this case we might want to flag up the option for users to run `nix develop` in place of installing `direnv` as I can imagine people being concerned about security.

**Comment by phated** [comment]
> After this PR we're going to be funnelling people down route B but with direnv we can automatically inject the flake's build environment into the user's main shell, making it roughly equivalent to the shell provided by nix-shell. Is that correct?

I'm not actually familiar with `nix-shell` but it seems like it is the same as `nix develop`. Instead of making people enter a `nix develop` environment, it felt better to rely on direnv which uses `nix print-dev-env` to populate the environment. This means that devs have the natural feel of their terminal instead of starship (or some arbitrary shell we choose).

My other reason for not recommending `nix develop` is due to it only working with one workflow: entering a new shell via your terminal and then using `code .` to open the editor with that exact environment. However, it doesn't work with some other workflows, such as:
1. Opening a new vscode folder via the `File` menu
2. Launching a project via WSL
3. Launching a project on a remote machine
4. Possibly even launching a devcontainer project (but I haven't explored this at all)

It's easiest for us to recommend the `code .` workflow in our docs since they are already in a terminal to run the other commands, but enough devs will be using the other workflows that I'd like to make sure our recommended steps solve for most/all.

I was thinking about adding another section to the docs for useful Nix commands, so we can probably add `nix develop` there. How does that sound?

**Comment by TomAFrench** [comment]
Ah yeah, I meant `nix develop` when I said `nix-shell`. Agreed that direnv is a better experience but I just wanted to make sure I understood what was going on ("nix develop but in your main shell") before I signed off on it.

I took a look at the new changes and they look perfect.

**Review by TomAFrench — APPROVED** [review]

**Comment by phated** [comment]
Koby and I are still working through problems with the ssh connections, so we'll wait on this for a bit.

File: README.md [hunk 1]
```diff
@@ -40,6 +40,80 @@ pub fn pedersen() {
...
+
+### Building against a different local/remote version of Barretenberg
+
+If you are working on this crate, it is likely that you want to incorporate changes from some other version of Barretenberg
+instead of the version this project is pinned against.
+
+To reference a different version of Barretenberg, you can add the `--override-input` flag. For example:
+
+```sh
+nix build . --override-input barretenberg /home/phated/barretenberg
```
Comment (vezenovm) [inline README.md:102]: This notes that I can build against a different version of Barretenberg than the one this project is pinned against. Under the Dependencies section it says that `barretenberg` needs to built and installed following the instructions in the bberg README. While running this command locally it looks to build barretenberg on its own. Do I need to follow step 4 of Dependencies if I am building through nix?

Comment (phated) [reply README.md:102]: The top of the README is primarily for the benefit of publishing to crates.io, where it explains the native dependencies that you'd need if you were integrating into a rust project.

We then hide all of that complexity with the Nix package manager for devs working on anything in the `noir-lang` organization. The docs at the bottom, such as this, are for folks working directly on Noir projects.

File: README.md [hunk 2]
```diff
@@ -40,6 +44,82 @@ pub fn pedersen() {
...
+```sh
+code .
+```
+6. (Recommended) When launching VSCode for the first time, you should be prompted to install our recommended plugins. We highly recommend installing these for the best development experience.
+
+### Building and testing
+
+Assuming you are using `direnv` to populate your environment, building and testing the project can be done
+with the typical `cargo build`, `cargo test`, and `cargo clippy` commands. You'll notice that the `cargo` version matches the
+version we specify in [flake.nix](./flake.nix), which is 1.66.0 at the time of this writing.
```
Comment (TomAFrench) [inline README.md:94]:
```suggestion
with the typical `cargo build`, `cargo test`, and `cargo clippy` commands. You'll notice that the `cargo` version matches the version we specify in [flake.nix](./flake.nix), which is 1.66.0 at the time of this writing.
```

**Comment by kevaundray** [comment]
After [448cc90](https://github.com/noir-lang/barretenberg-sys/pull/15/commits/448cc90d3c1763f74deadb47763928c2e69f170f) I tested that it works on linux.

Before we were getting a clang linker error problem, but Blaine fixed it

**Review by kevaundray — APPROVED** [review]

**Pull request merged**